File: erty-mobile/lib/domain/entities/budget.dart
import 'package:equatable/equatable.dart';

enum BudgetPeriod { daily, weekly, monthly, yearly }

class Budget extends Equatable {
  final String id;
  final String userId;
  final String? categoryId;
  final String name;
  final double amount;
  final BudgetPeriod period;
  final DateTime startDate;
  final DateTime endDate;
  final double spent;
  final int alertThreshold;
  final bool alertSent;
  final String color;

  const Budget({
    required this.id,
    required this.userId,
    this.categoryId,
    required this.name,
    required this.amount,
    required this.period,
    required this.startDate,
    required this.endDate,
    this.spent = 0.0,
    this.alertThreshold = 80,
    this.alertSent = false,
    required this.color,
  });

  double get percentage => (spent / amount) * 100;
  double get remaining => amount - spent;
  bool get isExceeded => spent > amount;
  bool get isNearLimit => percentage >= alertThreshold;

  @override
  List<Object?> get props => [
        id,
        userId,
        categoryId,
        name,
        amount,
        period,
        startDate,
        endDate,
        spent,
        alertThreshold,
        alertSent,
        color,
      ];

  Budget copyWith({
    String? id,
    String? userId,
    String? categoryId,
    String? name,
    double? amount,
    BudgetPeriod? period,
    DateTime? startDate,
    DateTime? endDate,
    double? spent,
    int? alertThreshold,
    bool? alertSent,
    String? color,
  }) {
    return Budget(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      categoryId: categoryId ?? this.categoryId,
      name: name ?? this.name,
      amount: amount ?? this.amount,
      period: period ?? this.period,
      startDate: startDate ?? this.startDate,
      endDate: endDate ?? this.endDate,
      spent: spent ?? this.spent,
      alertThreshold: alertThreshold ?? this.alertThreshold,
      alertSent: alertSent ?? this.alertSent,
      color: color ?? this.color,
    );
  }
}

File: erty-backend/src/models/Budget.js
const { DataTypes } = require('sequelize');
const { sequelize } = require('../config/database');

const Budget = sequelize.define('Budget', {
  id: {
    type: DataTypes.UUID,
    defaultValue: DataTypes.UUIDV4,
    primaryKey: true
  },
  userId: {
    type: DataTypes.UUID,
    allowNull: false,
    references: {
      model: 'users',
      key: 'id'
    }
  },
  categoryId: {
    type: DataTypes.UUID,
    allowNull: true, // null = orçamento geral
    references: {
      model: 'categories',
      key: 'id'
    }
  },
  name: {
    type: DataTypes.STRING,
    allowNull: false
  },
  amount: {
    type: DataTypes.DECIMAL(15, 2),
    allowNull: false,
    validate: {
      min: 0
    }
  },
  period: {
    type: DataTypes.ENUM('daily', 'weekly', 'monthly', 'yearly'),
    defaultValue: 'monthly'
  },
  startDate: {
    type: DataTypes.DATE,
    allowNull: false
  },
  endDate: {
    type: DataTypes.DATE,
    allowNull: false
  },
  spent: {
    type: DataTypes.DECIMAL(15, 2),
    defaultValue: 0
  },
  alertThreshold: {
    type: DataTypes.INTEGER,
    defaultValue: 80, // Alerta quando atingir 80%
    validate: {
      min: 0,
      max: 100
    }
  },
  alertSent: {
    type: DataTypes.BOOLEAN,
    defaultValue: false
  },
  isActive: {
    type: DataTypes.BOOLEAN,
    defaultValue: true
  },
  rollover: {
    type: DataTypes.BOOLEAN,
    defaultValue: false // Transferir saldo não gasto para próximo período
  },
  color: {
    type: DataTypes.STRING,
    defaultValue: '#4CAF50'
  },
  notes: {
    type: DataTypes.TEXT
  }
});

module.exports = Budget;
